lib/charms/grafana_k8s/v0/grafana_metadata.py

```diff
@@ -0,0 +1,250 @@
+
+Return:
+        self.log(f"Got Grafana's internal_url: {metadata.internal_url}")
+```
+
+### Provider
+
+To add this relation to your charm as a provider, add the following to your `charmcraft.yaml` or `metadata.yaml`:
+
+```yaml
+provides:
+  grafana-metadata:
+    interface: grafana_metadata
+```
+
+To manage sending data to related applications in your charm, use `GrafanaMetadataProvider`.  Note that
+`GrafanaMetadataProvider` *does not* manage any events, but instead provides a `send_data` method for sending data to
+all related applications.  Triggering `send_data` appropriately is left to the charm author, although generally you want
+to do this at least during relation_joined and leader_elected events.  An example implementation is:
+
+```python
+class FooCharm(CharmBase):
+    def __init__(self, framework):
+        super().__init__(framework)
+        self.grafana_metadata = GrafanaMetadataProvider(
+            charm=self,
+            grafana_uid=self.unique_name,
+            ingress_url=self.external_url,
+            internal_url=self.internal_url,
+            relation_name="grafana-metadata"
+        )
+
+        self.framework.observe(self.on.leader_elected, self.do_something_to_send_data)
+        self.framework.observe(self._charm.on["grafana-metadata"].relation_joined, self.do_something_to_send_data)
+```
+"""
+import json
+import logging
+from typing import List, Optional, Union
+
+
+from ops import CharmBase, BoundEvent, EventBase, CharmEvents, EventSource, Object
+from pydantic import BaseModel, Field
+
+# The unique Charmhub library identifier, never change it
+LIBID = "26290f24974540adb4464b695bd01ea3"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["pydantic>=2"]
+
+log = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "grafana-info"
+
+
+class GrafanaMetadataAppData(BaseModel):
+    """Data model for the grafana-info interface."""
+
+    ingress_url: str = Field(description="The ingress URL.")
+    internal_url: str = Field(description="The URL for connecting to the prometheus api from inside the cluster.")
+    grafana_uid: str = Field(description="The UID of this Grafana application.")
+
+
+class DataChangedEvent(EventBase):
+    """Charm Event triggered when the relation data has changed."""
+
+
+class ReceiverCharmEvents(CharmEvents):
+    """Events raised by the data receiver side of the interface."""
+
+    data_changed = EventSource(DataChangedEvent)
+
+
+class GrafanaMetadataRequirer(Object):
+    """Class for handling the receiver side of the grafana-info relation."""
+
+    on = ReceiverCharmEvents()  # type: ignore[reportAssignmentType]
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str,
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
```

Comment:
I regret the moment this ever became a pattern. IMO this is a better signature
```suggestion
        refresh_events: Sequence[BoundEvent] = (),
```

Reply:
hahaha I agree with you, but went for consistency with the others.  If others are onboard too I'm happy to change

Reply:
So are we agreed we can go with `refresh_events` (which seems more correct, but is inconsistent with some past libs like [grafana-source](https://github.com/canonical/grafana-k8s-operator/blob/13fa90ed7d5304ce26eaba17f0dff3c2831c07a8/lib/charms/grafana_k8s/v0/grafana_source.py#L326)?

As for the type in @PietroPasotti's suggestion, I think his suggestion looks good.  The type we accept is a Sequence (list, mutable or not, is fine for our purposes) but then we default to an empty tuple (which is a sequence).  @sed-i wdyt?